bazel_tools/build_environment.bzl

```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2020 The DAML Authors. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+def _impl(ctx):
+    # Generates an empty BUILD file, because we do not need to build anything.
+    ctx.file(
+        "BUILD",
+        content = "",
+        executable = False,
+    )
+
+    # Generates a simple Bazel file that just sets a bunch of Bazel variables,
+    # so they can be used in our main Bazel BUILD files.
+    semver = ctx.os.environ.get("DAML_SDK_RELEASE_VERSION", default = "0.0.0")
```

Comment:
> Just to make sure I understand the implications of this correctly:
> 
> * Everything that currently depends on `//:VERSION` and embeds it in some form will default to embedding `0.0.0` instead.

Yes.

> * If I want to reproduce a release commit build, it is not sufficient to checkout the commit. I have to set the environment variable.

Yes.

> * In a release build everything that embeds the version number will get the the version number specified in `LATEST`.

With the exception of Maven artifacts, which get a +100, and some GHC-internal things which (for snapshot builds) get a slightly simplified version string.

This is not magic though, this is CI setting the env var based on the LATEST file of the release commit (the one triggering the release) while building the source commit (the one whose SHA is in the LATEST file in the release commit).

So if LATEST says `abc 0.1.2`, CI is doing

```
git checkout abc
DAML_SDK_RELEASE_VERSION=0.1.2 bazel build //...
```

> * `daml-sdk-head` will embed `0.0.0` whereas before it embedded things from `LATEST`.

`daml-sdk-head always had `0.0.0`. It used to take artifacts built with `VERSION` and then "manually" patching them to change their version number before embedding them in the tarball. Now it builds the tarball directly from artifacts built as `0.0.0`.

> My two main worries here are:
> 
> 1. Caching will get significantly worse for release builds.

Release builds were already specifically builds that changed the VERSION file, so this is not getting worse in any way that I can see.

> 2. If there are issues relating to the version number (I’m particularly thinking of `damlc`, `daml-stdlib` and cross-SDK dependencies here. I’m not _aware_ of anything that should go wrong but this is finicky), we will not catch them until we do a release build which actually changes the version number used in the build.

True, and thanks for raising. I will change the CI configuration to run those tests against the release PR (rather than not run them at all, which is what this PR currently does).

> 
> I don’t have a great answer for either of those. I’d feel a bit more comfortable about 2 if we ran tests on release builds with the version number set but that also means release builds will basically rebuild and retest the world (I believe both sandbox and damlc embed the version number). Still seems better than catching the issue due to manual testing.

I agree. Though I strongly suspect that if our tests break on a version change like that, it's going to be a false positive, i.e. a wrong assumption in the test rather than an error in the binary produced.

> 
> The only other idea that I have is to continue for now to embed the SDK version from `LATEST` if the environment variable is not set. That at least seems to reduce the number of moving parts compared to the current SDK release process and should help with 1 and partially with 2 although it doesn’t quite match up.

I don't think that solves anything, really. It would not change anything to 1; the only difference for 2 is that currently we do effectively run the test suite with the env var set on the release PR (so, again, adding this here would not make it worse, AFAICT).